UnicastPublishRouter no longer uses LINQ in SelectDestinationsForEachEndpoint

danielmarbach opened the pull request:
Stumbled over the UnicastPublishRouter over the weekend and did a mental exercise to see if I could refactor it. The outcome is a slightly more complex code which seems to have less memory footprint and is overall much faster. Would we want this despite the complexity?

Scenarios used for benchmarking

https://github.com/danielmarbach/MicroBenchmarks/blob/master/MicroBenchmarks/NServiceBus/UnicastPublishRouter.cs#L26

``` ini

BenchmarkDotNet=v0.10.1, OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4980HQ CPU 2.80GHz, ProcessorCount=8
Frequency=2728067 Hz, Resolution=366.5599 ns, Timer=TSC
  [Host]     : Clr 4.0.30319.42000, 32bit LegacyJIT-v4.6.1586.0
  DefaultJob : Clr 4.0.30319.42000, 32bit LegacyJIT-v4.6.1586.0


```
 Method |                                    Scenario |          Mean |     StdErr |      StdDev |        Median | Scaled | Scaled-StdDev |  Gen 0 | Allocated |
------- |-------------------------------------------- |-------------- |----------- |------------ |-------------- |------- |-------------- |------- |---------- |
 **Before** |                                       **Empty** |   **222.7757 ns** |  **0.5123 ns** |   **1.9843 ns** |   **223.2220 ns** |   **1.00** |          **0.00** | **0.0369** |     **260 B** |
  After |                                       Empty |    81.5929 ns |  0.1433 ns |   0.5552 ns |    81.5410 ns |   0.37 |          0.00 | 0.0130 |      84 B |
 **Before** |                                **NullEndpoint** |   **406.7690 ns** |  **2.7216 ns** |  **10.5408 ns** |   **401.2914 ns** |   **1.00** |          **0.00** | **0.0621** |     **444 B** |
  After |                                NullEndpoint |   149.1392 ns |  0.2411 ns |   0.9337 ns |   148.9896 ns |   0.37 |          0.01 | 0.0279 |     180 B |
 **Before** |               **NullEndpointAndNormalEndpoint** |   **752.7702 ns** |  **2.8666 ns** |  **10.7257 ns** |   **749.9985 ns** |   **1.00** |          **0.00** | **0.0786** |     **608 B** |
  After |               NullEndpointAndNormalEndpoint |   465.4413 ns |  3.1289 ns |  12.1182 ns |   470.9883 ns |   0.62 |          0.02 | 0.0585 |     392 B |
 **Before** | **NullEndpointAndNormalEndpointWithRedundancy** | **1,098.3925 ns** | **24.3387 ns** | **100.3509 ns** | **1,073.9841 ns** |   **1.00** |          **0.00** | **0.0817** |     **652 B** |
  After | NullEndpointAndNormalEndpointWithRedundancy |   533.9369 ns |  0.5457 ns |   1.9676 ns |   534.1861 ns |   0.49 |          0.04 | 0.0551 |     396 B |
 **Before** |                    **NullEndpointAndTwoOthers** | **2,175.7344 ns** | **21.4606 ns** |  **95.9749 ns** | **2,135.1933 ns** |   **1.00** |          **0.00** | **0.1205** |    **1.1 kB** |
  After |                    NullEndpointAndTwoOthers | 1,289.5935 ns | 16.5135 ns | 115.5942 ns | 1,240.0484 ns |   0.59 |          0.06 | 0.0443 |     696 B |

timbussmann commented:
given that the numbers show quite an improvement, I'm good with pulling this in 👍

danielmarbach commented:
@timbussmann I realised that the before part was not materialising the second select. So it was an unfair comparison. I updated the test, let it run again and updated the PR description. The difference is no even bigger

timbussmann commented:
@bording feel free to merge in case you also think that these numbers are worth the given code.
